### ISSUE TITLE
Bug: Does not support recursive types

### ISSUE THREAD

**Issue by andykais** [woutervh-/typescript-is#12]
Recursive types do not compile, and throw a `RangeError: Maximum call stack size exceeded`. Let me know if this is a problem with `ttypescript` and I can report this bug upstream. Thank you for working on this library! 😄

stack trace:
```bash
$ npm run compile:runtime:config

/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:87527
                throw e;
                ^
NestedError: Failed to transform node at: /Users/andrew/Code/scratchwork/scrape-pages/src/settings/config/assert.ts:9:55
    at transformNodeAndChildren (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/transformer.js:30:15)
    at ts.visitEachChild (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/transformer.js:32:62)
    at visitNode (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:63887:23)
    at Object.visitEachChild (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:64161:59)
    at transformNodeAndChildren (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/transformer.js:32:15)
    at ts.visitEachChild (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/transformer.js:32:62)
    at visitNodes (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:63938:48)
    at Object.visitEachChild (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:64157:45)
    at transformNodeAndChildren (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/transformer.js:32:15)
    at ts.visitEachChild (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/transformer.js:32:62)
Caused By: RangeError: Maximum call stack size exceeded
    at Object.isNodeKind (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:13880:24)
    at Object.createNode (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:16206:22)
    at createSynthesizedNode (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:59514:23)
    at Object.createTypeOf (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript/lib/typescript.js:60561:20)
    at visitString (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/visitor.js:470:131)
    at visitType (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/visitor.js:514:16)
    at createPropertyCheck (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/visitor.js:12:24)
    at visitPropertySignature (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/visitor.js:51:12)
    at visitDeclaration (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/visitor.js:55:16)
    at visitPropertySymbol (/Users/andrew/Code/scratchwork/scrape-pages/node_modules/typescript-is/lib/transform-inline/visitor.js:111:16)
```

simplified failing snippet:
```ts
import { assertType } from 'typescript-is'
type ConfigInit = {
  folder: string
  children: ConfigInit
}

export const assertConfigType = (configInit: any) => {
  assertType<ConfigInit>(configInit)
}
```

tsconfig.json:
```json
{
  "compilerOptions": {
    "rootDir": "../../../../src",
    "outDir": "./",
    "sourceMap": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "lib": ["es2015", "es2016", "es2017"],
    "module": "commonjs",
    "moduleResolution": "node",
    "target": "es2017",
    "allowJs": true,
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "plugins": [
      {
        "transform": "typescript-is/lib/transform-inline/transformer"
      }
    ]
  },
  "include": ["src/**/*.ts", "custom.d.ts"],
  "files": ["../assert.ts", "../../../../custom.d.ts"]
}
```

**Comment by woutervh-** [comment]
Hi @andykais 

Good catch, recursive types are expanded indefinitely. I will make it a priority to support those as well.

Thanks!

**Comment by woutervh-** [comment]
hi @andykais 

I've finished a large refactoring and as part of the consequences of the refactor recursive types are now supported.
Please try it out using vesion `0.10.0` and let me know if it's working or if you have some other feedback.

Thanks!

**Comment by andykais** [comment]
😲 you work fast! I see they work on your example, but I used your assertions against my slightly more complex type ([ConfigInit](https://github.com/andykais/scrape-pages/blob/master/src/settings/config/types.ts)) and it seems to consider any object as valid. I played around a little, and I believe the issue is when there are optional fields on a type, the recursive checks no longer work.

I created a pr which fails a test to describe the behavior simply. Let me know if you need more info, and thanks again!

**Comment by woutervh-** [comment]
hi @andykais 

Again, good catch! There was a mistake in the code generation related to optional objects.
I have fixed it and released it as version `0.10.1`.
Please try again.

Thanks!

**Comment by andykais** [comment]
that did the trick! I have officially moved my project over to typescript-is! https://github.com/andykais/scrape-pages/commit/03f9ddf0fa5df3d9da40715c93e76173050a9dfe

Also, nice work improving those error messages! They are much clearer now! My problem has been addressed, so I will close this issue now
